server/src/main/java/org/elasticsearch/index/mapper/CustomSyntheticSourceFieldLookup.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.index.IndexSettings;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Contains lookup information needed to perform custom synthetic source logic.
+ * For example fields that use fallback synthetic source implementation or fields that preserve array ordering
+ * in synthetic source;
+ */
+public class CustomSyntheticSourceFieldLookup {
+    private final Map<String, Reason> fieldsWithCustomSyntheticSourceHandling;
+
+    public CustomSyntheticSourceFieldLookup(Mapping mapping, IndexSettings indexSettings, boolean isSourceSynthetic) {
+        this.fieldsWithCustomSyntheticSourceHandling = new HashMap<>();
```

Comment:
Maybe assign this after populating the map and then calling `ImmutableMap.copyOf`, so that no more updates can happen after construction.